syntax = "proto3";
import "google/protobuf/struct.proto";
import "common.proto";

package proto.ext.ui;

enum ToastStyle {
  Success = 0;
  Info = 1;
  Warning = 2;
  Error = 3;
  Dynamic = 4;
};

message ShowToastRequest {
  string id = 1;
  string title = 2;
  string message = 3;
  ToastStyle style = 4;
};

message HideToastRequest {
  string id = 1;
};

message UpdateToastRequest {
  string id = 1;
  string title = 2;
};

message PushViewRequest {}
message PopViewRequest {}
message CloseMainWindowRequest {
  bool clear_root_search = 1;
  PopToRootType pop_to_root = 2;
};
message ClearSearchBarRequest {};

message SetSearchTextRequest {
  string text = 1;
};

message GetSelectedTextRequest {
}

message GetSelectedTextResponse {
  string text = 1;
}

enum PopToRootType {
  PopToRootDefault = 0;
  PopToRootImmediate = 1;
  PopToRootSuspended = 2;
};

message ShowHudRequest {
  string text = 1;
  bool clear_root_search = 2;
  PopToRootType pop_to_root = 3;
};

message RenderRequest {
   //repeated RenderNode views = 1;
   // we will migrate to actual protobuf
   string json = 1;
};

enum ConfirmAlertActionStyle {
  Default = 0;
  Destructive = 1;
  Cancel = 2;
};

message ConfirmAlertRequest {
  string title = 1;
  string description = 2;
  Image icon = 3;
  ConfirmAlertAction dismiss_action = 4;
  ConfirmAlertAction primary_action = 5;
  bool remember_user_choice = 6;
};

message ConfirmAlertResponse {
  bool confirmed = 1;
};

message ConfirmAlertAction {
  string title = 1;
  ConfirmAlertActionStyle style = 2;
};

message PopToRootRequest {
  bool clear_search_bar = 1;
};

message Request {
  oneof payload {
    RenderRequest render = 1;
    ShowToastRequest show_toast = 2;
    HideToastRequest hide_toast = 3;
    UpdateToastRequest update_toast = 4;
    PushViewRequest push_view = 5;
    PopViewRequest pop_view = 6;
    ClearSearchBarRequest clear_search = 7;
    CloseMainWindowRequest close_main_window = 8;
    ShowHudRequest show_hud = 9;
    SetSearchTextRequest set_search_text = 10;
    ConfirmAlertRequest confirm_alert = 11;
    GetSelectedTextRequest get_selected_text = 12;
    PopToRootRequest pop_to_root = 13;
  };
};


message Response {
  oneof payload {
    common.AckResponse render = 1;
    common.AckResponse show_toast = 2;
    common.AckResponse hide_toast = 3;
    common.AckResponse update_toast = 4;
    common.AckResponse push_view = 5;
    common.AckResponse pop_view = 6;
    common.AckResponse clear_search = 7;
    common.AckResponse close_main_window = 8;
    common.AckResponse show_hud = 9;
    common.AckResponse set_search_text = 10;
    ConfirmAlertResponse confirm_alert = 11;
    GetSelectedTextResponse get_selected_text = 12;
    common.AckResponse pop_to_root = 13;
  };
};

message RenderNode {
  string type = 1;
  bool has_dirty_child = 2;
  bool has_dirty_props = 3;
  map<string, google.protobuf.Value> props = 4;
  repeated RenderNode children = 5;
};

message ThemedImageSource {
  string light = 1;
  string dark = 2;
};

message ImageSource {
  oneof payload {
    string raw = 1;
    ThemedImageSource themed = 2;
  };
};

enum ImageMask {
  None = 0;
  Circle = 1;
  RoundedRectangle = 2;
};

message DynamicColor {
  string light = 1;
  string dark = 2;
  optional bool adjust_contrast = 3;
};

message ColorLike {
  oneof payload {
    string raw = 1;
    DynamicColor dynamic = 2;
  };
};

message Image {
  ImageSource source = 1;
  optional ImageSource fallback = 2;
  optional ImageMask mask = 3;
  optional ColorLike tint_color = 4;
};
